client/transaction-pool/src/lib.rs

```diff
@@ -727,6 +787,139 @@ where
 	}
 }
 
+/// Helper struct for deciding if core part of maintenance procedure shall be executed.
+/// It is publicly exposed only for tests purposes (due to cyclic deps sc_transaction_pool and
+/// substrate_test_runtime_transaction_pool).
+///
+/// For the following chain:
+///   B1-C1-D1-E1
+///  /
+/// A
+///  \
+///   B2-C2-D2-E2
+///
+/// Some scenarios and expected behavior:
+/// nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(C1))
+/// f(C1), nbb(C1) -> false (handle_enactment was already performed in f(C1))
+///
+/// f(C1), nbb(D2) -> false (handle_enactment was already performed in f(C1), we should not retract
+/// finalized block)
+/// f(C1), f(C2), nbb(C1) -> false
+/// nbb(C1), nbb(C2) -> true (switching fork is OK)
+/// nbb(B1), nbb(B2) -> true
+/// nbb(B1), nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(B1)
+/// nbb(C1), f(B1) -> false (handle_enactment was already performed in nbb(B2)
+#[doc(hidden)]
+pub struct EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	recent_best_block: Option<Block::Hash>,
+	recent_finalized_block: Option<Block::Hash>,
```

Comment:
We should initialize this at startup to `finalized_hash` and `best_block_hash`. Then we can get rid off the `Option`. This will make the code much easier.

Reply:
This requires further extension of `ChainApi` with `info()` method, correct?

Reply:
[3793021](https://github.com/paritytech/substrate/pull/12305/commits/379302165e47176faf4f89a67e3d858c7085ee6d)